File: 이상욱/BOJ/m3w6/BOJ1074.java
package m3w11;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;


public class BOJ1074 {
	static int answer;
	public static void main(String[] args) throws Exception{
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		
		int N = Integer.parseInt(st.nextToken());
		int r = Integer.parseInt(st.nextToken());
		int c = Integer.parseInt(st.nextToken());
		
		int mapSize = (int) Math.pow(2, N);
		answer = 0;
		
		checkZ(mapSize, r, c);
		System.out.println(answer);
	}
	private static void checkZ(int mapSize, int r, int c) {
		
		if(mapSize == 1)
			return;
		
		if(r < mapSize/2 && c < mapSize/2) {
			checkZ(mapSize/2, r, c);
		}
		else if(r < mapSize/2 && c >= mapSize/2) {
			answer += mapSize * mapSize / 4;
			checkZ(mapSize/2, r, c - mapSize/2);
		}
		else if(r >= mapSize/2 && c < mapSize/2) {
			answer += (mapSize * mapSize / 4) * 2;
			checkZ(mapSize/2, r - mapSize/2, c);
		}
		else {
			answer += (mapSize * mapSize / 4) * 3;
			checkZ(mapSize/2, r - mapSize/2, c - mapSize/2);
		}
		
	}
}

File: 이상욱/BOJ/m3w6/BOJ1074.kt
package m3w11

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

var answer = 0
fun main(){

    val br = BufferedReader(InputStreamReader(System.`in`))
    val st = StringTokenizer(br.readLine())

    val N = st.nextToken().toInt()
    val r = st.nextToken().toInt()
    val c = st.nextToken().toInt()

    val mapSize = (Math.pow(2.0, N.toDouble())).toInt()
    
    checkZ(mapSize,r,c)
    println(answer)

}

fun checkZ(mapSize: Int, r: Int, c: Int) {

    if(mapSize == 1){
        return
    }
    if(r < mapSize / 2 && c < mapSize / 2){
        checkZ(mapSize / 2,r,c)
    }else if(mapSize / 2 in (r + 1)..c){
        answer += mapSize * mapSize / 4
        checkZ(mapSize / 2 , r, c - mapSize/2)
    }else if(mapSize/2 in (c + 1)..r){
        answer += (mapSize * mapSize / 4) * 2
        checkZ(mapSize / 2, r - mapSize / 2, c)
    }else{
        answer += (mapSize * mapSize) / 4 * 3
        checkZ(mapSize / 2, r - mapSize / 2 , c - mapSize/2)
    }

}
